src/txmempool.cpp

```diff
@@ -505,17 +980,79 @@ bool CCoinsViewMemPool::GetCoins(const uint256 &txid, CCoins &coins) const {
+    if (mempool.lookup(txid, cert)) {
+        LogPrint("cert", "%s():%d - making coins for cert [%s]\n", __func__, __LINE__, txid.ToString() );
+        coins = CCoins(cert, MEMPOOL_HEIGHT);
+        return true;
+    }
     return (base->GetCoins(txid, coins) && !coins.IsPruned());
 }
 
 bool CCoinsViewMemPool::HaveCoins(const uint256 &txid) const {
     return mempool.exists(txid) || base->HaveCoins(txid);
 }
 
+bool CCoinsViewMemPool::GetSidechain(const uint256& scId, CSidechain& info) const {
+    if (mempool.hasSidechainCreationTx(scId)) {
+        //build sidechain from txs in mempool
+        const uint256& scCreationHash = mempool.mapSidechains.at(scId).scCreationTxHash;
+        const CTransaction & scCreationTx = mempool.mapTx.at(scCreationHash).GetTx();
+        for (const auto& scCreation : scCreationTx.GetVscCcOut()) {
+            if (scId == scCreation.scId) {
+                //info.creationBlockHash doesn't exist here!
+                info.creationBlockHeight = -1; //default null value for creationBlockHeight
+                info.creationTxHash = scCreationHash;
+                info.creationData.withdrawalEpochLength = scCreation.withdrawalEpochLength;
+            }
+            break;
+        }
+
+    } else if (!base->GetSidechain(scId, info))
+        return false;
+
+    //decorate sidechain with fwds and bwt in mempool
+    if (mempool.mapSidechains.count(scId)) {
+        for (const auto& fwdHash: mempool.mapSidechains.at(scId).fwdTransfersSet) {
+            const CTransaction & fwdTx = mempool.mapTx.at(fwdHash).GetTx();
+            for (const auto& fwdAmount : fwdTx.GetVftCcOut())
+                if (scId == fwdAmount.scId)
+                    info.mImmatureAmounts[-1] += fwdAmount.nValue;
+        }
+
+        if (!mempool.mapSidechains.at(scId).backwardCertificate.IsNull()) {
+            const uint256& certHash = mempool.mapSidechains.at(scId).backwardCertificate;
+            const CScCertificate & cert = mempool.mapCertificate.at(certHash).GetCertificate();
+            info.balance -= cert.GetValueOfBackwardTransfers();
+        }
+    }
+
+    return true;
+}
+
+bool CCoinsViewMemPool::HaveSidechain(const uint256& scId) const {
+    return mempool.hasSidechainCreationTx(scId) || base->HaveSidechain(scId);
+}
+
+bool CCoinsViewMemPool::HaveCertForEpoch(const uint256& scId, int epochNumber) const {
+    if ((mempool.mapSidechains.count(scId) != 0) && (!mempool.mapSidechains.at(scId).backwardCertificate.IsNull()))
+        return true;
+
+    return base->HaveCertForEpoch(scId, epochNumber);
+}
+
 size_t CTxMemPool::DynamicMemoryUsage() const {
     LOCK(cs);
-    return memusage::DynamicUsage(mapTx) + memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) + cachedInnerUsage;
+    return
+        ( memusage::DynamicUsage(mapTx) +
+          memusage::DynamicUsage(mapNextTx) +
+          memusage::DynamicUsage(mapDeltas) +
+          memusage::DynamicUsage(mapCertificate) +
```

Comment:
I guess, that `mapSidechains` must be included as well.

Reply:
Done